File: unity/Assets/Scripts/Core/DragonGateJudge.cs
namespace DragonGateSlot.Core
{
    public enum JudgeResult { Pass, HitWall, Miss, SameValueHit }

    public struct RowJudgment
    {
        public int Row;
        public JudgeResult Result;
        public int Gap; // hi - lo, meaningful only for Pass
        public bool HasScatter; // any cell in this row is Scatter
    }

    public static class DragonGateJudge
    {
        /// <summary>Judge all 3 rows. Returns array of 3 RowJudgments.</summary>
        public static RowJudgment[] JudgeBoard(Cell[,] board)
        {
            var results = new RowJudgment[3];
            for (int r = 0; r < 3; r++)
                results[r] = JudgeRow(r, board[r, 0], board[r, 1], board[r, 2]);
            return results;
        }

        static RowJudgment JudgeRow(int row, Cell left, Cell mid, Cell right)
        {
            var j = new RowJudgment { Row = row };
            j.HasScatter = left.Type == CellType.Scatter ||
                           mid.Type == CellType.Scatter ||
                           right.Type == CellType.Scatter;

            // If any cell is Scatter, this row is not judged (treated as miss)
            if (j.HasScatter)
            {
                j.Result = JudgeResult.Miss;
                return j;
            }

            int l = left.Value, m = mid.Value, r2 = right.Value;

            // Same-value gate
            if (l == r2)
            {
                j.Result = (m == l) ? JudgeResult.SameValueHit : JudgeResult.Miss;
                return j;
            }

            // Normal gate
            int lo = l < r2 ? l : r2;
            int hi = l > r2 ? l : r2;

            if (m == lo || m == hi)
            {
                j.Result = JudgeResult.HitWall;
            }
            else if (m > lo && m < hi)
            {
                j.Result = JudgeResult.Pass;
                j.Gap = hi - lo;
            }
            else
            {
                j.Result = JudgeResult.Miss;
            }
            return j;
        }
    }
}

File: web/js/dragonGateJudge.js
export const JudgeResult = { Pass: 'Pass', HitWall: 'HitWall', Miss: 'Miss', SameValueHit: 'SameValueHit' };

export function judgeBoard(board) {
  const judgments = [0, 1, 2].map(r => judgeRow(r, board[r][0], board[r][1], board[r][2]));
  let scatterCount = 0;
  for (let r = 0; r < 3; r++)
    for (let c = 0; c < 3; c++)
      if (board[r][c].isScatter) scatterCount++;
  return { judgments, scatterCount };
}

function judgeRow(row, left, mid, right) {
  const j = { row, result: JudgeResult.Miss, gap: 0 };
  const l = left.value, m = mid.value, r2 = right.value;

  // Same-value gate
  if (l === r2) {
    j.result = (m === l) ? JudgeResult.SameValueHit : JudgeResult.Miss;
    return j;
  }

  // Normal gate
  const lo = Math.min(l, r2);
  const hi = Math.max(l, r2);

  if (m === lo || m === hi) {
    j.result = JudgeResult.HitWall;
  } else if (m > lo && m < hi) {
    j.result = JudgeResult.Pass;
    j.gap = hi - lo - 1;
  }
  return j;
}
